### PR TITLE
SG-38522 support note reply syncing

### PR DESCRIPTION
**Pull request by pscadding**
Adds Reply syncing to the Jira bridge. 
Jira has comments and we currently sync Notes <> Comments. On Jira Cloud the Comments can have parent ids associated for creating a threaded conversation (restricted to 1 level). This matches Flow PT's Note and Reply pattern nicely, so this feature now syncs Flow PT replies to Jira comments with parent ids and the reverse.

Since comment replies in Jira default to mentioning the original user, I've also added some basic handling of the mentions on the Flow PT side. They still look a little bit ugly since Flow PT doesn't support mentions but does remap the Jira user ID to the Flow PT user so its a little more readable. This remapping is handled both ways when the mentioned users are linked in Flow PT to Jira Users. See the two bracketed parts in the screenshot below for an example of how Jira mentions looks in Flow PT.
Screenshot 2026-09-03 at 12 33 37
When the mentioned Jira user is not mapped to a Flow PT user, then the mention is not reformatted and preserves the raw text syntax Jira uses, when syncing to Flow PT.

Note that the user can't create fields on Replies in Flow PT, so this requires a field to be created on the Note that stores the mapping of all child replies to Jira comments.

### CODE REVIEW

File: settings.py [hunk 1]
```diff
@@ -192,6 +192,10 @@
                     "sg_entity": "Note",    # Note is a special entity, we only need to add the "sg_entity" key if we want to sync the changes
                     "sync_deletion_direction": "both_way",
                 },
+                {
+                    "sg_entity": "Reply",   # Reply is synced as Jira comment replies on the parent Note's comment
+                    "sync_deletion_direction": "both_way",
```
Comment (pscadding) [inline settings.py]: We don't do it else where, but I suppose we could have a setting for allowing deletion, on one or either sides.

File: settings.py [hunk 2]
```diff
@@ -192,6 +192,10 @@
...
+                },
```
Comment (pscadding) [inline settings.py]: I've opted to put this as a separate entry. I think this makes sense, since maybe some are still using a version of Jira without replies, and also the Reply entity requires a new field created on the Note.

My only reason for thinking that it should not be separate is that is it so tightly coupled with Note, that if you enabled Note you would nearly always want Reply as well.

I think Reply should be added as default to the settings as well. We just need to test what happens when the field is not created.

File: sg_jira/bridge.py [hunk 3]
```diff
@@ -333,10 +333,15 @@ def get_syncer(self, name):
...
+            # Only cache the syncer once it's successfully set up - if setup()
+            # raises (e.g. a required custom field is missing), the syncer must
+            # not be cached, so the next attempt retries setup() and fails the
+            # same clear way, instead of silently skipping validation forever.
+            syncer = syncer_class(
                 name=name, bridge=self, hook_class=hook_class, **settings
             )
-            self._syncers[name].setup()
+            syncer.setup()
+            self._syncers[name] = syncer
```
Comment (pscadding) [inline sg_jira/bridge.py:344]: This isn't strictly required, I added it because as I was going through the test scenarios as written in the ticket, It meant that we only got the expected error once, and then proceeded to get runtime errors for each occurrence/sync attempt after that.

File: azure-pipelines/tests.yml [hunk 4]
```diff
@@ -61,7 +61,7 @@ jobs:
           --cov-report xml:coverage.xml \
           --durations=0 \
           --nunit-xml=test-results.xml \
-          --verbose \
+          --verbose
```
Comment (pscadding) [inline azure-pipelines/tests.yml:64]: So I'm a little unsure on this. The CI was failing for me until these changes were made (with the suggestion of AI). However, this didn't seem to be causing issues when this was implemented earlier this year, and on other PRs.

File: sg_jira/handlers/entities_generic_handler.py [hunk 5]
```diff
@@ -440,6 +570,16 @@ def accept_jira_event(self, resource_type, resource_id, event):
...
+            if not sync_settings:
+                # Unlike Issue-type events, comment/reply/worklog events have no
+                # sync_settings of their own to fall back on if the FPTR entity
+                # type isn't configured - falling through to the parent Issue's
+                # settings further down would wrongly accept the event.
+                self._logger.debug(
+                    f"Rejecting Jira event: FPTR {sg_entity_type} is not configured "
+                    f"for syncing."
+                )
+                return False
```
Comment (pscadding) [inline sg_jira/handlers/entities_generic_handler.py:582]: This is fixing a pre-existing bug. Prior to this a full sync of an issue could wrongly sync a child note/comment even if the bridge was not configured to do so.

File: settings.py [hunk 6]
```diff
@@ -191,6 +191,10 @@
                 {
                     "sg_entity": "Note",    # Note is a special entity, we only need to add the "sg_entity" key if we want to sync the changes
                     "sync_deletion_direction": "both_way",
+                    # Replies are synced as Jira comment replies on this Note's comment.
+                    # Reply has no settings of its own - it fully inherits sync_direction
+                    # and sync_deletion_direction from this entry.
+                    "enable_reply_syncing": True,
```
Comment (pscadding) [inline settings.py:197]: As discussed with Barbara, we felt that Replies shouldn't be a separate entity listed in the sync settings from Notes. I did originally have it this way, but it added room for increased complexity that is just not needed. For example you could have different sync direction settings for Notes compared to Replies and things start to get complicated with edge cases. 
Really replies should just be handled the same as Notes, so moving replies to be a setting of the Note syncing felt like a better approach.

File: azure-pipelines/tests.yml (hunk 4 above)
Comment (pscadding) [reply azure-pipelines/tests.yml:64]: UPDATE: I was comparing to Kuldeep's PR when I said it didn't impact other PR's and it seems he has needed to make the same change in order for CI to pass as well (which I hadn't spotted).

File: sg_jira/handlers/entities_generic_handler.py [hunk 7]
```diff
@@ -1667,6 +1892,106 @@ def _sync_jira_entity_to_sg(
...
+            self._logger.debug(
+                f"Couldn't find FPTR Note with Jira key {jira_note_key}. Skipping Jira reply sync."
+            )
+            return False
+
+        raw_mapping = sg_note.get(SHOTGUN_JIRA_REPLY_IDS_FIELD) or ""
+        try:
+            mapping = json.loads(raw_mapping) if raw_mapping else {}
+        except (ValueError, TypeError):
+            mapping = {}
```
Comment (barbara-darkshot) [inline sg_jira/handlers/entities_generic_handler.py:1929]: ### sg_jira_reply_ids handling when the mapping JSON is invalid or manually edited

In _sync_jira_reply_to_sg (and __get_reply_mapping on the FPTR→Jira path), we fall back to {} when json.loads fails. That keeps sync from blocking on corrupt data, but it also means the next write replaces the whole field with only the entry for the current operation — any previously valid mappings in that string are effectively discarded.

Worth being explicit about the two cases:

Fully invalid JSON (e.g. "not-json") → mapping is reset; existing FPTR Replies / Jira replies are not deleted, but the links are lost and we risk duplicates on the next sync.
Valid JSON with a bad/missing entry → other entries are kept; only that reply’s sync is affected (may recreate or skip).
There’s no general reconciliation that rebuilds the mapping from all Jira comment children + FPTR Replies — backfill loops help only in specific paths and don’t fully repair a broken mapping.

Not blocking, but flagging in case manual edits to this field are a realistic ops scenario. Happy to keep the current “fail open” behavior, or we could consider alternatives (e.g. refuse to overwrite on parse failure, log/error more prominently, or a repair pass from Jira children before writing)

File: sg_jira/hook.py [hunk 8]
```diff
@@ -328,13 +353,80 @@ def get_sg_user_from_jira_user(self, jira_user):
...
+        def _replace(match):
+            sg_user = sg_users_by_account_id.get(match.group(1))
+            if not sg_user:
+                return match.group(0)
+            name = re.sub(r"\s+", "", sg_user["name"])
+            return "[mention:%s:%s]" % (sg_user["id"], name)
+
+        return self.JIRA_MENTION_REGEX.sub(_replace, body)
+
+    def _rewrite_sg_mentions_to_jira(self, body):
```
Comment (barbara-darkshot) [inline sg_jira/hook.py:389]: do we want to have "private" method here? or should we use public method if the goal is to be overridden by users (if needed)
(same for`_rewrite_jira_mentions_to_sg`)

File: tests/python/test_sync_base.py [hunk 9]
```diff
@@ -58,5 +58,30 @@ def setUp(self):
 
         self.mock_jira_session_bases()
 
+        # Patch mockgun.Shotgun with no-op stubs for real SG API methods that
+        # mockgun doesn't implement (add_user_agent added in SG-42067,
+        # set_session_uuid called by bridge.sync_in_jira).
+        for _method in ("add_user_agent", "set_session_uuid"):
```
Comment (barbara-darkshot) [inline tests/python/test_sync_base.py:64]: I guess you've made these modifications in order to avoid doing the modifications directly in the python API?
maybe we can at least create a Jira ticket to track these changes and make sure they could be done at the root level itself at some point?

**Review by barbara-darkshot — APPROVED** [review]
LGTM, only added a few small comments!

File: sg_jira/handlers/entities_generic_handler.py (hunk 7 above)
Comment (pscadding) [reply sg_jira/handlers/entities_generic_handler.py:1929]: Yeah I think this is very edge case. The sort of thing that I don't expect to happen in the wild, but if it was a problem then we could address it when raised.

Since data shouldn't be lost, only potentially duplicated its not the worst thing.

File: sg_jira/hook.py (hunk 8 above)
Comment (pscadding) [reply sg_jira/hook.py:389]: It should still be overrideable right?
The main intention here of marking it as private was that it would not be called outside of the hook or a descendant hook.
